bindings/go/oci/transformer/extract.go

```diff
@@ -0,0 +1,183 @@
+		return nil, fmt.Errorf("failed to read OCI layout: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, store.Close())
+	}()
+
+	mainArtifacts := store.MainArtifacts(ctx)
+	if len(mainArtifacts) != 1 {
+		return nil, fmt.Errorf("should have exactly one main artifact but was %d", len(mainArtifacts))
+	}
+
+	// Parse configuration
+	var extractConfig *spec.Config
+	if config != nil {
+		if cfg, ok := config.(*spec.Config); ok {
+			extractConfig = cfg
+		}
+	}
+
+	artifact := mainArtifacts[0]
+	return t.extractOCIArtifact(ctx, store, artifact, extractConfig)
+}
+
+// extractOCIArtifact extracts selected layers from an OCI artifact into a tar archive.
+func (t *Transformer) extractOCIArtifact(ctx context.Context, store content.Fetcher, artifact ociImageSpecV1.Descriptor, config *spec.Config) (_ blob.ReadOnlyBlob, err error) {
+	manifestReader, err := store.Fetch(ctx, artifact)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch artifact manifest: %w", err)
+	}
+	defer func() {
+		err = errors.Join(err, manifestReader.Close())
+	}()
+
+	manifestData, err := io.ReadAll(manifestReader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read manifest data: %w", err)
+	}
+
+	var manifest ociImageSpecV1.Manifest
+	if err := json.Unmarshal(manifestData, &manifest); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal manifest: %w", err)
+	}
+
+	var tarBuffer bytes.Buffer
+	tarWriter := tar.NewWriter(&tarBuffer)
+	defer func() {
+		err = errors.Join(err, tarWriter.Close())
+	}()
+
+	// If no config provided, extract all layers with default naming
+	if config == nil || len(config.Rules) == 0 {
+		for _, layer := range manifest.Layers {
+			filename := t.getDefaultFilename(layer.Digest.String())
+			if err := t.processLayer(ctx, store, layer, tarWriter, filename); err != nil {
+				return nil, fmt.Errorf("failed to process layer %s: %w", layer.Digest, err)
+			}
+		}
+	} else {
+		// Process layers according to configured rules
+		for _, rule := range config.Rules {
+			if err := t.processRule(ctx, store, manifest.Layers, rule, tarWriter); err != nil {
+				return nil, fmt.Errorf("failed to process rule for file %s: %w", rule.Filename, err)
+			}
+		}
+	}
+
+	resultBlob := inmemory.New(bytes.NewReader(tarBuffer.Bytes()))
+	resultBlob.SetMediaType("application/tar")
+
+	return resultBlob, nil
+}
+
+// processRule processes layers that match the rule's selectors into a single tar file.
+func (t *Transformer) processRule(ctx context.Context, store content.Fetcher, layers []ociImageSpecV1.Descriptor, rule spec.Rule, tarWriter *tar.Writer) error {
+	for i, layer := range layers {
+		layerInfo := spec.LayerInfo{
+			Index:       i,
+			MediaType:   layer.MediaType,
+			Annotations: layer.Annotations,
+		}
+
+		// Check if this layer matches any of the rule's selectors
+		matched := false
+		for _, selector := range rule.LayerSelectors {
+			if selector.Matches(layerInfo) {
+				matched = true
+				break
+			}
+		}
+
+		if matched {
```

Comment:
nit: switch this if and continue and then move the code out of the if block to make nesting smaller